File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/utils/dependencies/PluginDependenciesNode.java
package com.jetbrains.pluginverifier.utils.dependencies;

import com.intellij.structure.domain.Plugin;
import com.intellij.structure.domain.PluginDependency;
import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * @author Sergey Patrikeev
 */
public final class PluginDependenciesNode {
  @NotNull
  private final Plugin myPlugin;

  /**
   * All plugins reachable from myPlugin.
   */
  @NotNull
  private final Set<Plugin> myTransitiveDependencies;

  /**
   * Missing dependency -> reason why it is missing.
   */
  @NotNull
  private final Map<PluginDependency, MissingReason> myMissingDependencies;

  /**
   * The set of existing {@link PluginDependenciesNode}s reachable from {@code this} node.
   */
  @NotNull
  private final Set<PluginDependenciesNode> myEdges;

  PluginDependenciesNode(@NotNull Plugin plugin,
                         @NotNull Set<PluginDependenciesNode> edges,
                         @NotNull Set<Plugin> transitiveDependencies,
                         @NotNull Map<PluginDependency, MissingReason> missingDependencies) {
    myPlugin = plugin;
    myTransitiveDependencies = new HashSet<>(transitiveDependencies);
    myMissingDependencies = new HashMap<>(missingDependencies);
    myEdges = new HashSet<>(edges);
  }

  @NotNull
  public Set<PluginDependenciesNode> getEdges() {
    return Collections.unmodifiableSet(myEdges);
  }

  @NotNull
  public Set<Plugin> getTransitiveDependencies() {
    return Collections.unmodifiableSet(myTransitiveDependencies);
  }

  @NotNull
  public Map<PluginDependency, MissingReason> getMissingDependencies() {
    return Collections.unmodifiableMap(myMissingDependencies);
  }

  @NotNull
  public Plugin getPlugin() {
    return myPlugin;
  }

  @Override
  public String toString() {
    return myPlugin.getPluginId();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    PluginDependenciesNode that = (PluginDependenciesNode) o;

    return myPlugin.getPluginFile().equals(that.myPlugin.getPluginFile());
  }

  @Override
  public int hashCode() {
    return myPlugin.getPluginFile().hashCode();
  }
}

File: verifier-cli/src/main/kotlin/com/jetbrains/pluginverifier/utils/dependencies/PluginDependenciesNode.kt
package com.jetbrains.pluginverifier.utils.dependencies

import com.intellij.structure.domain.Plugin
import com.intellij.structure.domain.PluginDependency

/**
 * @author Sergey Patrikeev
 */
class PluginDependenciesNode(val plugin: Plugin,
                             /**
                              * The set of existing [PluginDependenciesNode]s reachable from `this` node.
                              */
                             val edges: Set<PluginDependenciesNode>,
                             /**
                              * All plugins reachable from [plugin].
                              */
                             val transitiveDependencies: Set<Plugin>,
                             /**
                              * Missing dependency -> reason why it is missing.
                              */
                             val missingDependencies: Map<PluginDependency, MissingReason>) {


  override fun toString(): String {
    return plugin.pluginId
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false

    val that = other as PluginDependenciesNode?

    return plugin.pluginFile == that!!.plugin.pluginFile
  }

  override fun hashCode(): Int {
    return plugin.pluginFile.hashCode()
  }
}
